jax/_src/api_util.py

```diff
@@ -132,6 +134,96 @@ def __hash__(self):
   def __eq__(self, other):
     return type(self.val) is type(other.val) and self.val == other.val
 
+_POSITIONAL_ARGUMENTS = (
+  inspect.Parameter.POSITIONAL_ONLY,
+  inspect.Parameter.POSITIONAL_OR_KEYWORD
+)
+
+def validate_argnums(sig: inspect.Signature, argnums: Tuple[int, ...], argnums_name: str) -> None:
+  """
+  Validate that the argnums are sensible for a given function.
+
+  For functions that accept a variable number of positions arguments
+  (`f(..., *args)`) all positive argnums are considered valid.
+  """
+  n_pos_args = 0
+  for param in sig.parameters.values():
+    if param.kind in _POSITIONAL_ARGUMENTS:
+      n_pos_args += 1
+
+    elif param.kind is inspect.Parameter.VAR_POSITIONAL:
+      # We can have any number of positional arguments
+      return
+
+  if (
+    -min(argnums, default=1) >= n_pos_args
```

Comment:
should this be `-min(argnums, default=1) > n_pos_args`?
IIUC, `-n_pos_args` is the first argument.

Reply:
And, I personally think that
```python
if not all(-n_pos_args <= argnum < n_pos_args for argnum in argnums):
```
looks better than
```python
if -min(argnums, default=1) > n_pos_args or max(argnums, default=-1) >= n_pos_args:
```

Reply:
I agree with your first comment.

With regards to the second comment: I don't like the introduction of a loop, even if this could be argued to be more readable. Further, I don't actually think it is more readable. To me, the original code reads "Are we below lower bound or above upper bound", which IMO reads quite nicely.

TODO
- [ ] Fix bound check (first comment)

Reply:
Agreed, my suggestions is somewhat nit-pick. -- Another point: considering #10669 , is it safe to raise an error right now? @hawkinsp

Reply:
The short-circuit if check to get around default is quite neat. Definitely like that!

Reply:
I don't quite follow the question about #10669.

Reply:
Changed over to short-circuit if statement:
`if argnums and (-min(argnums) > n_pos_args or max(argnums) >= n_pos_args)`